app/src/main/java/org/stepic/droid/ui/activities/SmartLockActivityBase.kt

```diff
@@ -0,0 +1,146 @@
+                            resolvingWasShown = true
+                            credentialRequestResult.status.startResolutionForResult(this, REQUEST_FROM_SMART_LOCK_CODE)
+                        }
+                    } catch (e: IntentSender.SendIntentException) {
+                        Timber.e(e, "STATUS: Failed to send resolution.")
+                    }
+
+                } else {
+                    Timber.d("STATUS: Failed to send resolution.")
+                    // The user must create an account or sign in manually.
+                }
+            }
+        }
+    }
+
+    protected fun requestToSaveCredentials(authData: AuthData) {
+        val credential = Credential
+                .Builder(authData.login)
+                .setPassword(authData.password)
+                .build()
+
+        Auth.CredentialsApi.save(googleApiClient, credential)
+                .setResultCallback { status ->
+                    if (!status.isSuccess && status.hasResolution()) {
+                        analytic.reportEvent(Analytic.SmartLock.SHOW_SAVE_LOGIN)
+                        status.startResolutionForResult(this, REQUEST_SAVE_TO_SMART_LOCK_CODE)
+                    } else {
+                        analytic.reportEventWithName(Analytic.SmartLock.DISABLED_LOGIN, status.statusMessage)
+                        onCredentialSaved()
+                    }
+                }
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        when (requestCode) {
+            REQUEST_FROM_SMART_LOCK_CODE -> {
+                if (resultCode == RESULT_OK && data != null) {
+                    analytic.reportEvent(Analytic.SmartLock.LAUNCH_CREDENTIAL_RETRIEVED_PROMPT)
+                    val credential = data.getParcelableExtra<Credential>(Credential.EXTRA_KEY)
+                    onCredentialRetrieved(credential)
+                } else {
+                    analytic.reportEvent(Analytic.SmartLock.LAUNCH_CREDENTIAL_CANCELED_PROMPT)
+                    Timber.d("Credential Read not ok: canceled or no internet")
+                }
+            }
+
+            REQUEST_SAVE_TO_SMART_LOCK_CODE -> {
+                if (resultCode == RESULT_OK) {
+                    analytic.reportEvent(Analytic.SmartLock.LOGIN_SAVED)
+                } else {
+                    analytic.reportEvent(Analytic.SmartLock.LOGIN_NOT_SAVED)
+                }
+                onCredentialSaved()
+            }
+        }
+
+        super.onActivityResult(requestCode, resultCode, data)
+    }
+
+    override fun onSaveInstanceState(outState: Bundle?) {
+        outState?.putBoolean(RESOLVING_ACCOUNT_KEY, resolvingWasShown)
+        super.onSaveInstanceState(outState)
+    }
+
+    protected fun signOutFromGoogle() {
+        if (googleApiClient?.isConnected == true) {
+            Auth.GoogleSignInApi.signOut(googleApiClient)
+        }
+    }
+
+    protected open fun onCredentialRetrieved(credential: Credential) {}
```

Comment:
let's make it abstract

Reply:
If we make it abstract classes that don't need this functionality will have to have empty implementation of this method.